### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by barneycarroll** [joelrich/citojs#18]
Cito looks very nice. There are some interesting approaches here — the use of fragments and the special diffing policy for inputs strike me as very neat features.

I'm interested in what you'd like to do next, and what you think the challenges are. I've been using Mithril for the past year and have seen components and lifecycle hooks produce all sorts of interesting problems. Cito has yet to make any mistakes in this department, so I'd be keen to hear your thoughts on these issues :)

**Comment by joelrich** [comment]
Thanks for the compliment :-) I am still working on a rewrite that looks similar to Cito, but works differently and should perform even better in real case scenarios. Unfortunately, it takes a little bit longer than expected to implement it.

Please let me know if you have some concrete requirements for the lifecycle events and if it fits with my plans, then I will try to implement them.

**Comment by barneycarroll** [comment]
That sounds interesting! 

What do you mean by 'looks similar to Cito'? Coming from a year with Mithril, Cito syntax seems very similar to the output of m() functions — in fact earlier today I was toying with the idea of writing a function to compile m() signature to Cito virtual DOM syntax output. I think this is one of the best things about Mithril — the flexible elegance and convenience of being able to write large dynamic DOM templates succinctly.

>…works differently…

Intriguing. What do you want to change about it? It's already the most efficient DOM abstraction differ algorithm going by the benchmarks…!

> concrete requirements for the lifecycle events

This is higher-order stuff that reaches in 2 directions: upwards to stateful components, and downwards to low-level DOM access. AFAICT Cito deals exclusively with DOM abstraction ATM so how to approach these issues is very interesting for me. One of the things I believe Mithril has failed at is providing good hooks similar to React's `didMount`/`willUnmount` etc. More importantly, I believe both React and Mithril have shortcomings in the way they implement this kind of stuff, usually involving verbose and intuitive APIs that compromise the benefits of efficient and flexible virtual DOM by only enabling these APIs with direct relationships between virtual DOM and references to specific real DOM nodes. For example, a unique mobile component or virtual DOM `config` in Mithril mandates creating a new DOM structure — this sacrifices most of the benefits of a 'virtual' DOM. At the other end of the spectrum there's Famo.us, which was ultra efficient in its control over DOM abstraction, but required learning a completely new DSL. I think somewhere in between is desirable, and this can be done by considering the failings of React + Mithril's constructive pragmatism compared to Famo.us' naive ambition.

TLDR: Nothing specific, I'm just curious and excited and have a lot of opinions :) _but_ I am thinking of writing an (ultra-inefficient) extension of Mithril that will provide the higher-order lifecycle APIs I've been missing. This would basically be a refined combination of [flexible component management](https://gist.github.com/barneycarroll/0e3c1b9811b47c012b13), [element lifecycle hooks](https://gist.github.com/barneycarroll/c2dda75bcc94a7424d75), and [DOM transformers](https://github.com/barneycarroll/mattr).

**Comment by joelrich** [comment]
When I implemented Cito, I never liked the fact that it has to create a whole virtual dom just to figure out a few changes. Most templates are pretty much static. Why should anyone want to create and diff the whole virtual dom then?

I have thought a lot about how this could be solved and have already implemented a few prototypes. There are still a few open questions, but in general I like my idea quite and will try to publish a first version of it soon.

By the way, I implemented lifecycle events a while ago. However, there are still a few issues with them and I haven't documented them yet. I tried to solve the problem by providing the following three events:

```
events: {
	$created: function (e) {
		// e.type, e.target 
	},
	$changed: function (e) {
		// e.type, e.target, e.changes
	},
	$destroyed: function (e) {
		// e.type, e.target
	}
}
```

That looks quite similar to your init, draw and exit hook. Although I have wanted to, I have never implemented support for a context object.

**Comment by barneycarroll** [comment]
I'm not sure that having the lifecycle hooks in the events hash makes that much sense. IMO it would make more sense to have these in their own hash:

```Javascript 
{
	tag : 'input',
	attrs : {
		type : 'search'
	},
	events : {
		input : e =>
			model.searchString = e.target.value
	},
	lifecycle : {
		init : el =>
			el.focus()
	}
}
```

DOM event objects aren't that nice. One of the advantages of virtual DOM abstractions is that they provide the possibility of a cleaner syntax than native DOM, with all its bloat and confusion. `type` isn't useful in this syntax, and exposing relevant arguments directly to the callback function would be far more useful. Making the lifecycle hooks look like DOM event handlers leaves the door open for all sorts of unhelpful cognitive dissonance: Can I trigger `$destroyed`? Can I `preventDefault`on a `$changed` event? etc. Breaking these out to their own distinct API would also make implementation logic far cleaner: no more string-matching conditional logic in the events iteration loop.

Furthermore, I believe Mithril and React have suffered in their reliance in too close and vague a binding between virtual and real DOM. Making the distinction clear offers clarity for users and avoids potential for conflicts of interest in advanced usage.

WRT `context`, my experience of using Mithril for a year has revealed that `context` is technically redundant and arguably an anti-pattern. If you need access to stateful stores in an immutable view, what you really want is component state. Mithril needed `context` because component controllers were too cumbersome to initialise and reason about. The real challenge is to make components easy and intuitive. If a virtual DOM structure (even if that structure is a single node) needs it's own state, then the user should have one go to method for doing that: initialising it as a sub-component.

WRT granular redraws, I think this should be a component-based thing too. The Mithril community has been hankering for this for a while — there's a pull request for that pending.

**Comment by gaydenko** [comment]
Sometimes, I think vast majority of developers developing front-end frameworks are mysterious:) Let's start from the beginning.

What are we doing? We are mastering an SPA. The application has it's own intention and logic. The logic can be expressed different ways. Say, the way suggested by Nicholas Zakas (and I know pretty implementation of it: http://www.scaleapp.org/ ). That is an application probably contains something called modules with: parent/children, init/destroy, start/stop, pub/sub, sandboxes and so on. Some of these modules can deal with UI, that is with DOM. Well, let's repeat: *some* of them.

In any UI context (and especially dealing with virtual DOM) we eventually deal with tree of subtrees (let's call the last ones components). We want to manipulate DOM (via virtual DOM as an appealing way) from application modules: to insert and to remove components (probably inside already existing components), to direct events and data flows, and so on.

The main magic lays here: many front-end application frameworks are designed the way when components tree and application structure are considered to be almost the same (or be very closely correlated). And this is a point I don't understand :) And all these frameworks are an attempt to smooth that outstanding differences between application structure and components tree. Aren't they?

Let's separate these structures. Let's application modules be application, and let's components be something handy manipulated from any application (sub)modules.

All the talk written above (being really humble opinion) is a reason I"m looking for a way to easy manipulate UI fragments (regardless DOM parent-child hierarchy) from a "detached" point (that is from modules). 

As well as @barneycarroll (if I'm not wrong), I feel cito.js is (at the moment) free from being application framework (being UI-layer framework only). So I'm also interested in the upcoming cito.js state, and, especially, in your, @joelrich , intention to keep cito.js application-logic free. Also, will API and sense be mostly kept to start to convert mid-size application to cito.js today? :)

**Comment by barneycarroll** [comment]
@gaydenko I like what you're saying (I think — personally I find it's easy to misunderstand each-other when talking about high level programming architecture ;). I hadb't seen scaleapp, I'll be sure to read into it and watch the talk.

From what I gather you are talking about a fallacy in web app development whereby people link DOM tree structure too closely to model hierarchy structure. In most simple straw man applications this is incidentally accurate and convenient, but at scale it becomes an insidious anti-pattern. One of the reasons I have high hopes for cito is that Mithril has made a lot of mistakes in developing an API that is over-reliant on a direct relationship between application state and real DOM. It's a sad irony, because the most empowering aspect of the virtual DOM paradigm is the liberation from this fallacy (see [this expose by James Long](http://jlongster.com/Removing-User-Interface-Complexity,-or-Why-React-is-Awesome). I've made myself quite infamous in the Mithril community by repeatedly warming against this anti-pattern, which is why I think cito is in a good position to avoid the mistakes of previous attempts.

One of the big difficulties in the front end world is that most front end developers are very bad at reasoning about models and statefulness: we come from an environment where statefulness is provided by servers and the document object model, and for a long time we were just extending or enhancing these mechanisms with flourishes of script — it's still in very recent history where web development with JavaScript was referred to as 'dynamic HTML'. But I digress ;)

There's a bit of ambiguity in your statement, in that it isn't clear whether you think cito should be aiming to address these problems in a progressive approach to 'componentization' and / or state machines, or leave this problem domain as an 'outside context' and stick to being a pure view library and rely on other tools to be used in conjunction for the issue of statefulness. What do you think?

**Comment by gaydenko** [comment]
@barneycarroll ,
> you are talking about a fallacy in web app development whereby people link DOM tree structure too closely to model hierarchy structure

Not exactly. I'd say, an application aim is the first, and data model (be it business data or component's one) are the second.
> but at scale it becomes an insidious anti-pattern

I don't guess insidiousness was an intention :)

Yes, I have read Jame's article recently. And on the way to find UI layer framework I have also stumbled with Mithril.js, but lately have found it to be "too application" for me. So, we are in par here (but I haven't used and don't know Mithril.js as you do). It's funny two keepers of not too shared opinion have met on the cito.js square :)

> There's a bit of ambiguity in your statement, in that it isn't clear whether you think cito should be aiming to address these problems in a progressive approach to 'componentization' and / or state machines, or leave this problem domain as an 'outside context' and stick to being a pure view library and rely on other tools to be used in conjunction for the issue of statefulness. What do you think?

The last one. I don't say the Unix way is always the way, but at the context I'd be happy to see cito.js clean view library keeping for developers an opportunity to adapt concrete library using in accordance with particular development and developer's context.

**Comment by barneycarroll** [comment]
@gaydenko thanks for clarifying, this helps my thinking a lot.

I guess I have a bit of a problem because in my mind I believe that the nix approach is the correct one, but I still kinda want one framework to handle the whole thing :)

I'm going to think long and hard about this. Thanks for the inspiration!

**Comment by joelrich** [comment]
@barneycarroll I agree that DOM events are clumsy and there is no need to implement the lifecycle hooks like DOM events. I will think about this again.

Should a low-level virtual dom library even support components? I have thought about this for a while and think that this isn't the responsibility of the virtual dom library itself. However, a virtual dom library must provide lifecycle hooks and a way to attach state/context to a virtual node to allow implementing an efficient component framework on top of it.

Who knows, maybe I will implement a component framework at some point but so far I have had the impression that there are enough other developers who are interested in doing this. One of the nice things of not having a component framework in the core library is that there might even be some competition to implement the best component framework for it - and they would still all work nicely together.

**Comment by gaydenko** [comment]
@joelrich ,

> Should a low-level virtual dom library even support components? I have thought about this for a while and think that this isn't the responsibility of the virtual dom library itself. However, a virtual dom library must provide lifecycle hooks and a way to attach state/context to a virtual node to allow implementing an efficient component framework on top of it.

Exactly! :+1:

**Comment by jails** [comment]
@barneycarroll 
> most front end developers are very bad at reasoning about models and statefulness
I don't think they are necessarily "bad". Large front end apps are still a new paradigm. And i think it'll requires time to find a "consensus", i.e a flexible enough architecture which will fit the needs of all coders imo. And the fact that everday a new front end framework is out shows that the community is still searching the best way to go.

Even if I don't really like all "team branded" frameworks like React or Angular, I must admit that the popularity they gained had the merit put a cat among the pigeons ! And a lot of coders started to rethink their front end architecture because of them.

Personnaly I found [deku](https://github.com/dekujs/deku) one of the most interesting outsider. However it won't  stop me to create my own front-end framework since I think some part can be enhanced.

I must join you on the fact that a virtual dom library shouldn't support components (it's a different concern imo). However it should still be able to manages/updates a set of trees. I added this through a well known `mount()`/`unmount()` strategy in [dom-layer](https://github.com/crysalead-js/dom-layer) and it simplified greatly any kind fo component layer which will sit on top of it.

**Comment by gcanti** [comment]
@joelrich @gaydenko

> Should a low-level virtual dom library even support components? I have thought about this for a while and think that this isn't the responsibility of the virtual dom library itself. However, a virtual dom library must provide lifecycle hooks and a way to attach state/context to a virtual node to allow implementing an efficient component framework on top of it.

:+1: 
This lib looks very nice.

**Comment by erlend-sh** [comment]
You and @lhorie seem to be [borrowing ideas](https://github.com/lhorie/mithril.js/issues/712#issuecomment-121089692) from one another all the time. Wouldn't you be better off collaborating on the same library? (which should probably be mithril, since it's more established).

Just seems to me like there's a lot of time spent on developing overlapping features.

**Comment by gaydenko** [comment]
@erlend-sh , mitshril is more rather just presentation layer (props, own ajaxing, own promises, own routing on so on). Some developers want just clean presentation layer. But borrowing is great - if you have an idea an I have an idea, and we exchange them. we will have two ideas both :)

**Comment by trueadm** [comment]
@erlend-sh Joel has been working on a new architecture that is quite a bit faster than the current Cito – so it would be nice to see if @lhorie and @joelrich  could work together to some degree on how to best approach performance given their experience.

**Comment by tinchoz49** [comment]
I can't belive the perfomance of citojs in this benchmark: http://mathieuancelin.github.io/js-repaint-perfs/ is increible fast.
@joelrich what do you think about open a gitter room or an IRC channel to tell us about your ideas in your rewrite of citojs, i can't wait to see it.

**Comment by joelrich** [comment]
@erlend-sh it would make sense if I kept Cito as it is; however, the ideas which I am working on will not be compatible with Mithril. I am not familiar with Mithril's code base, but I guess it would need a lot of refactoring to reach the same performance. What I did with Cito is rather insane. Almost every single decision was made in favour of performance. Doing this in retrospect without rewriting the whole is difficult. On the other hand, improving the most critical parts would probably be good enough for most use cases.

@tinchoz49 that's a good idea. I have been experimenting too much in private with my rewrite and have therefore not gotten enough feedback yet. Since it will require some API changes, this is very important though. For now, I would like to continue the discussion here. Let me share the basic idea with you:

All virtual dom libraries have the same flaw - they create a lot of object garbage only to figure out that most of the virtual dom hasn't changed. If you look at real templates, they are usually pretty static. Unfortunately, virtual dom libraries do not differentiate between static and variable parts of the template. For example, if we have the following template

```{tag: 'div', children: 'text'}```

then it cannot know that it will never change and creates an object and compares it every time.

By changing the API a little bit, this can be avoided. Assume we have the following template:

```
function div(text) {
	return {tag: 'div', children: text};
}
```

If we executed this function just once, passed a special object as the text parameter and analysed the returned virtual dom tree, then we could compare only the text variable in the future to figure out whether or not the dom has to be updated. However, that means that dynamic expressions must not be executed immediately. Therefore, we have to wrap them with callbacks or pass the calculated values to the template function. 

Let me clarify this with an example: Let's assume we only want to show the first three characters of the text variable. Since the template function is only executed once during a "preparation" step, we have to wrap the expression with a lambda/callback. The callback is then executed when the template is used. For updates, it will first just check whether the text variable value has changed and only then execute the lambda and update the dom accordingly.

```
function div(text) {
	return {tag: 'div', children: text.use(t => t.substr(0, 3))};
}
```

Instead of this, we could also just use the template with the calculated value:

```
function div(text) {
	return {tag: 'div', children: text};
}

var fragment = cito.appendTo(document.body); // Append a placeholder to the DOM
setTimeout(function () {
	var randomText = Math.random().toString(); // Create a random string
	fragment.render(div, randomText.substr(0, 3)); // Insert or update the placeholder with the div template function and the abbreviated random string as the first argument
}, 1000);
```

This API allows even more crazy optimisations, but requires wrapping all dynamic expressions. From my point of view, a virtual dom library should be used in combination with something like t7/JSX. In this case, there won't be any difference in usage because the transpiler can do the job for you. However, if you use Cito directly, then you would have to write your code slightly differently.

My impression is that most people are interested in Cito because of its performance. With this change, the performance will improve to a level that is unlikely to be outdone significantly anytime soon. Nevertheless, the flexibility of a virtual dom library can be kept. Therefore, this is the right next step.

What do you think? I am really interested in any thoughts about this proposal.

**Comment by barneycarroll** [comment]
@joelrich I like this idea. One of the really neat incidental APIs in Mithril is `m.prop`, which allows you to pass around getter/setter functions. It's a really trivial function, and serves mostly to show how cool simple functions are and offer an easy way to do bi-directional binding. But in practice it allows you to indicate 'possibility of change'. This is one of the things I like about cito's willingness to consume functions: there's a deferred execution, which is a semantic indication that the returned value may change. It's interesting to see how this might be exploited for practical uses.

@gaydenko really helped clarify my thinking on what it is I really want out of cito. Mithril is an awesome library, but it's efforts to keep things small, terse, scalable, _and_ idiomatic for application development have been huge burdens.

I'm currently developing a couple of apps with cito as the view. I'm looking at developing a simple function that converts Mithril `m` signature to cito output. Something I'll be looking at developing in the near future is an idiomatic way of referencing intermediary model constructors within views such that we have an 'MVC' method for writing components for citojs with some kind of lifecycle contract. Basically, a way of writing modular, behavioral, view-centric code with cito as the DOM manager.

Currently, the panoply of `cito.vdom` methods are strange to me. I hope to come up with some more constructive opinions in a week or so :)

**Comment by tailhook** [comment]
@joelrich If you didn't see incremental [DOM], you should look at it: 
It's interesting from two points of view:
1. They are trying to eliminate a lot of garbage collection too
2. According to [benchmark] incremental dom is slower than cito (4x-5x on many tests)

[DOM]: https://github.com/google/incremental-dom
[benchmark]: http://vdom-benchmark.github.io/vdom-benchmark/

**Comment by joelrich** [comment]
@barneycarroll there was a reason for each `cito.vdom` method, but I think I have now a better solution:

```
cito.appendTo(parentNode: HTMLElement): Fragment
cito.insertBefore(nextNode: Node): Fragment

Fragment.render(template: function, ...arguments: any): void
Fragment.remove(): void
```

Whereas you have to use `create()`/`append()` first and `update()` later at the moment, you can then simply always use `fragment.render()` as I tried to show with my last example in my comment above.

An alternative to `render(template: function, ...arguments: any`) might be that you have to pass the template to a `template()` function as in the following example:

```
function div(text) {
	return {tag: 'div', children: text};
}

var divTemplate = cito.template(div);
cito.appendTo(document.body).render(divTemplate('Some text...'));
```

Versus:

```
function div(text) {
	return {tag: 'div', children: text};
}
cito.appendTo(document.body).render(div, 'Some text...');
```

There are some special cases where `cito.template` might be advantageous, but I have to think about it further. How do you like these ideas?

**Comment by joelrich** [comment]
@tailhook incremental dom is interesting, but the API is quite ugly since they only intend it a compilation target. Moreover, they made a mistake in my opinion: While it's true that it doesn't create that much garbage, all the template code has still to be executed for each update. This is different with my idea because of the template preparation/compilation step that should allow a lot of optimisations which are impossible to achieve with an incremental/less declarative API.

**Comment by tinchoz49** [comment]
@joelrich the idea about differentiate between static and variable parts of the template is great. Is going to improve the perfomance and "delete?" (my english is soo bad) a lot of work in the diff algorithm.
About the API of cito i belive that keep a small and simple virtual dom library is the way. Over that we can use as you sayed t7 | jsx. The only thing that i think is really necesary in this kind of libraries is an API for the lifecycle as you sayed too haha.
About the idea of provide a way for attach states to a virtual node i don't undertstand what did you mean. If you mean provide an API for create states with getters/setters i don't know if is really necesary in this level, maybe is better keep free for the customer choice. For example, i love the idea of use citojs with baobabjs to manage my states, then if i need update my template i can use the event implementation of baobab.

Again excuse my English haha!

**Comment by joelrich** [comment]
@tinchoz49 with state I mean values which are bound to a dom node and its lifecycle. When the node is attached, the state must be initialised. When the node is detached, the state must be destroyed. In between, the state can be changed through events. On top of it you can then implement stateful components, as for example input fields.

**Comment by tinchoz49** [comment]
ahh ok! my mistake then. For now i've nothing more to say. 
It's a great library and i believe has great future. Great work!

**Comment by mneumann** [comment]
@barneycarroll as you are using cito.js, may I ask what libraries you are using instead of mithrils ```m.request``` and ```m.route``` functionality? I was using cito.js together with page.js and superagent, and recently I started to like https://github.com/github/fetch. Btw, cito.js helped me a lot to understand how mithril actually works, as in the beginning it was too much magic :)

**Comment by barneycarroll** [comment]
@mneumann yeah, I'm using `fetch` for XHR. Page.js looks decent - thanks for the tip!

@joelrich I don't understand the value in having different APIs here. `appendTo` and `insertBefore` are lifted out of the native DOM API,which already provides all these methods. It seems pointless to partially emulate it. Rather, Cito's API should simply take care of building and patching.

A 'processed' virtual tree features `dom` properties that map the virtual node to a DOM counterpart. Once we have access to these, we can use whatever native DOM methods we want to insert it into the document, change the `dom` property, etc.

AFAIC you only need 2 APIs:

```javascript
// Construct DOM from virtual structure, return 'processed' structure
cito.build( virtualStructure )

// Attempt to patch new virtual structure onto old one, fall back to build, return processed structure
cito.patch( domNode/virtualStructure, newVirtualStructure )
```

For a Mithril-style continuous universal diff system, I might have application code like this:

```javascript
requestAnimationFrame( function render(){
  cito.patch( document.body, {
    children : [
      {
        tag      : 'h1',
        children : new Date()
      }
    ]
  } )
  
  requestAnimationFrame( render )
} )
```

Of course, the idea of having one big tree structure that defines your entire application view may seem too simplistic depending on how difficult you want things to be. In this case you can still do any amount of complicated engineering:

```javascript
const header = cito.build( {
  tag      : 'header',
  children : {
    tag      : 'h1',
    children : 'Hello'
  }
} )

document.body.appendChild( header.dom )

const nav = cito.build( {
  tag      : 'nav',
  children : {
    tag      : 'ul',
    children : [
      tag      : 'li'
      children : {
        tag      : 'a'
        attrs    : {
          className : 'active',
          href      : '/'
        },
        children : 'Home'
      },
      tag      : 'li'
      children : {
        tag      : 'a'
        attrs    : {
          href : '/faqs'
        },
        children : 'FAQs'
      }
    ]
  }
} )

document.body.appendChild( nav.dom )

window.onpopstate = function(){
  cito.patch( nav, { /* new content based on route change */ } )
}
```

**Comment by lhorie** [comment]
Hey guys. Interesting discussion going on here :)

For those not in the loop, I'm working on a rewrite of Mithril and taking inspiration from cito, which is a very nice project.

@joelrich the idea of separating the dynamic parts of the template from the static via callbacks does seem like a upcoming point of convergence for virtual dom libraries. Both Glimmer (Ember) and Blaze (Meteor) have been doing something along those lines as well, and I've been thinking about it too as one of the more straightforward of a few ideas I have been considering. You're right that with Mithril's `m()` syntax, it would be probably not be possible to support pre-compilable templates while maintaining backwards compatibility and that a preprocessor or templating language would be needed, be it JSX, t7, handlebars, etc - unless one of my other ideas works out to be not mathematically impossible :)

That idea is basically the same thing, but statically compiling the dynamic parts out of a plain declarative js AST via esprima, instead of doing it at runtime from a higher-level templating language. The idea is that code could still run normally as it does now by falling back to traditional vdom diffing, but it could be also be aggressively compiled to do away with nothing-changed-but-still-needs-to-scan-the-whole-tree-to-be-sure situations.

Another idea was to do what Vue does and track data array operations as they happen by monkeypatching its methods (which can then be used to avoid a O(n) scan of the array at diff time), and maybe use ES6 Proxies to track object changes.

Some of these ideas might potentially be too crazy (and the latter one at least is clearly outside the scope of cito), but I just wanted to bounce them off other people and see what they think.

WRT lifecycle methods, one thing that has been asked a couple of times for Mithril is the ability to handle pre-detach animations (or, in a nutshell, asynchronous detachments). Another somewhat obscure recurring problem has to do with bubbling and/or similarly-timed events (e.g. input + keyup) and maintaining model-view synchronization correct while avoiding double redraws (without writing a bunch of convoluted code in app space). Another feature that was requested (albeit only once) was the ability to move expensive-to-initialize elements to different locations in a vdom tree. 

@barneycarroll @mneumann I'm curious to hear about your experience w/ fetch. I've had reservations about it due to the lack of maturity in some corner cases (e.g. aborting requests), but it being an upcoming standard API obviously makes it a very compelling option.

**Comment by tailhook** [comment]
> lifecycle methods, one thing that has been asked a couple of times for Mithril is the ability to handle pre-detach animations (or, in a nutshell, asynchronous detachments)

I'm pretty sure I've seen a virtual dom library that done that. I don't remember which one. And I think that its very interesting thing to have out of the box without any ugly things on application side.

> Another somewhat obscure recurring problem has to do with bubbling and/or similarly-timed events (e.g. input + keyup) and maintaining model-view synchronization correct while avoiding double redraws

Isn't this works by just redrawing in requestAnimationFrame's callback? Seems easy.

> Another feature that was requested (albeit only once) was the ability to move expensive-to-initialize elements to different locations in a vdom tree.

I don't think browsers are so slow that you can't just redraw. Unless either (a) you need to recompute  some state in component along with dom, or (b) you need to do that in every frame of animation, or (c) you are drawing a complex SVG image with virtual dom. Then (a) is fixed by detaching component from DOM which is non-trivial for React but easy with cito. The (b) is probably fixed by refactoring an animation to move element visually instead of moving hierarchy. And (c) is fixed by not drawing the image with virtual dom (i.e. using <img>) or similarly to (b). There might be exceptions, but I don't think they are frequent enough.

**Comment by gaydenko** [comment]
> I'm pretty sure I've seen a virtual dom library that done that. I don't remember which one. And I think that its very interesting thing to have out of the box without any ugly things on application side.

Agree. [Trackira](https://github.com/trackira/trackira) already [has got it](https://github.com/trackira/trackira/blob/master/src/Element/prototype/destroy.js) (and I use it for "animate before death"), and [vidom](https://github.com/dfilatov/vidom) is [on the way](https://github.com/dfilatov/vidom/issues/11), I hope. To my taste the feature is an absolute must-have.

**Comment by lhorie** [comment]
> I'm pretty sure I've seen a virtual dom library that done that. I don't remember which one.

Yeah, that one is probably the most important of the three things I mentioned. The other ones are more obscure. I just brought them up because they could also potentially complicate architecture. 

> Isn't this works by just redrawing in requestAnimationFrame's callback

Most of the time yes, except when you have multiple different components trying to read from the DOM (e.g. offsetHeight) to do further computations. Then you either bite the bullet and draw too many times, or you have to write async code in app space to manage the redraws.

> I don't think browsers are so slow that you can't just redraw. Unless either (a) you need to recompute some state in component along with dom, or (b) you need to do that in every frame of animation, or (c) you are drawing a complex SVG image with virtual dom.

The use case that was brought up was a HighCharts component IIRC. For the record, I'm also on the camp of leaving big honking expensive things alone and change only CSS if possible.

**Comment by llambda** [comment]
Is it useful to have async rendering of components? E.g., say you have a file upload component that you want to use AMD to load that javascript asynchronously, maybe that component returns a promise of the view that than eventually renders in the view. Would it be useful? Is it possible now?

**Comment by trueadm** [comment]
@lhorie I've done a massive amount of work on vdom optimisations with @joelrich. Parts of this work have gone into Inferno and the re-write of Cito.

Firstly, having a render loop generate a complete virtual DOM on each pass is expensive. It's expensive for two reasons:
- it causes quite a bit of garbage collection
- it's generating a lot of static objects that don't ever need to be diff'd

If you check out Inferno's DBmonster for example, you'll see that diffing values rather than vdom gave a good 25-30% performance boost compared to Cito. It was
almost on the same level as Rich Harris's vanilla implemenation.

The most effecient way (that we found) was to simply make the render generate an array of values of things that will change. For example:

```javascript
t7`<div class="bar">${ foo }<span>${ bar }</span><ul class="list">${ things.map( thing => t7`<li class="item">${ thing }</li>` ) }</ul></div>`
```

This would render the vdom on the first pass, then on further passes, it would return:

```javascript
[foo, bar, things.map( thing => [ thing ] )]
```

That's massively less data to return and it results in GC times of almost nothing in comparison to vdom GC times.

ES6 proxies are extremely slow in Firefox, to same degree as Object.observe in Chrome and Vue's
monkeypatching of array methods is also far slower than simply navigating an array. You speak of O(n) in terms of array scanning, but the JIT compiler only achieves that if it can create a low-level array to do so. When you monkeypatch the array methods, you lose all that JIT power and you're left with something that never touches upon near-native speed.

The key is in simplicity and looking at the compiled V8 code that these frameworks generate. It's all about micro-optimisations and understanding what works and what doesn't. There's no point in even touching any ES6 features, as they're all massively slower than the equivilants in ES5. For example, WeakMaps, Tagged Templates, Arrow Functions and Classes all suffer terribly in performance. Promises perform well mind you, but they've been about for a while now and have propably had some TLC.

I'd like to see native immutable objects for arrays and objects. Furthermore, if they're optimised well enough, you should see the biggest gains with those
objects in relation to diffing not only the vdom but the values.

**Comment by lhorie** [comment]
@trueadm Thanks for commenting. That's valuable information to have.

**Comment by trueadm** [comment]
I already had many conversations on Skype with Joel tonight about this stuff. It's very low-level, but ultimately interesting. In the attempt to give the V8/Gecko compiler something more efficient to work with, it will pay us back with performance and memory usage far lower than any crazy abstraction at the application level. As long as we're not touching the DOM and working efficiently with our own abstraction of the DOM, we can best make changes when they happen. I made a quick test to show some structures and their performance. It's not very scientific but it gives a good indication:

file:///Volumes/StorageVol/Sites/www/vdom-benchmark-inferno/test2.html

For me, the optimised vdom method (what I'm using in Inferno) is around the same speed as values in arrays. Sometimes it gets JIT compiled slightly faster, especially on Safari and Chrome, but never on Firefox it seems. I also wrote an implementation using typed arrays, but it was a headache waiting to happen, even if it did get the fastest performance by about 20-30% compared to values in arrays.

I'm trying to find an optimal virtual DOM structure that provides close to performance of simple values in arrays whilst allowing us to have a very tiny virtual DOM replica to better work with, without having to work with fragments or dynamic template factories. Using t7 I can optimise the output to cache static values/nodes and boost performance in large applications that contain large amounts of static content.

**Comment by jails** [comment]
I am a bit skeptical about all of this. For static html fragments why not simply using the `innerHTML` property (something like `{ props: { innerHTML: '<span>Hello I'm static</span>' } `) instead of creating such static structure using a virtual dom object ? Moreover using a templating langage like JSX / t7 or any other built-in solution should be able to perform such "static html optimizations" automatically.

And if we consider that an html page is compouned by components, each component render their own virtual dom (inside a mount point), manage their states (variables) and mark themself as "dirty" when their states change. Then only dirty components will need to be rerendered to minimize virtual dom rebuild & updates.

Did I miss the optimisation you are talking about ?

**Comment by localvoid** [comment]
@barneycarroll 
>I don't understand the value in having different APIs here. appendTo and insertBefore are lifted out of the native DOM API,which already provides all these methods.

It is because rendering in IE is way much faster this way: https://github.com/localvoid/vdom-benchmark/issues/15#issuecomment-71689398

@jails 
>For static html fragments why not simply using the innerHTML property (something like { props: { innerHTML: '<span>Hello I'm static</span>' }) instead of creating such static structure using a virtual dom object ?

https://github.com/facebook/react/issues/3226

**Comment by trueadm** [comment]
I've made a Skype group to talk about this virtual DOM related stuff in detail. If you have Skype, please add me "trueadm" and I'll add you to the group. This includes @Bobris @sebmarkbage @anthonyshort Matt-Esch

**Comment by joelrich** [comment]
@jails if a whole subtree is static, you could use html strings, but obviously it's then not possible anymore to apply transformations on the virtual dom. On the other hand, with my idea it would still be possible to apply transformations and it works also if there are still variable parts in a mostly static the subtree. There are definitely other approaches to solves this problem, but for now I am trying it this way.

Marking components as dirty is definitely an option, but unless you use something like `setState`, it's not that straightforward to implement.

**Comment by joelrich** [comment]
@lhorie thanks for your input. I thought about parsing the JS as well and although it would have the advantage that the API stays very simple, I think it would be overkill.

For now I am trying to solve this problem with separating the static from the dynamic part by using functions and callbacks. Although it's working fine and is faster for real use cases, I haven't been able to reach the current Cito performance for the initial dom creation in the vdom benchmark.

The reason is that the overhead for a list item with only one or two elements is higher than for a virtual dom node. I am still hoping to find a solution for this though.

**Comment by jails** [comment]
Interesting ! Thanks for the head up @localvoid @joelrich

**Comment by barneycarroll** [comment]
@localvoid very interesting — thanks for the recondite info! Obviously this is hardcore implementation optimisation and should be handled under the hood — unless cito really wants to be a low level tool for consumption by higher level frameworks rather than a get-up-and-go application development tool. 

@lhorie I think you're onto something there. Things have changed a lot since you recommended MSX and sweet.js in your blog posts — Babel (and it's controversial adoption of JSX compilation as part of core) has really changed the landscape: designing for compilation is much more viable now than a year ago (tangent: I had an idea for using `Function.prototype.toString` and a special `cache` method to memoize parts of the view, but this is probably more trouble than it's worth given variable implementation).

@trueadm seeing as this topic has mutated into a discussion of state-of-play for virtual DOM performance, would you mind giving us a potted history of inferno.js?

**Comment by barneycarroll** [comment]
@lhorie regarding `fetch`, in current practice it's basically what I was driving at when I tentatively suggested alternative signature to `m.request` - a rational API for HTTP doesn't bog the user down with the verbose tedium of XHR and returns promises. It's somewhat icky to use now because some aspects of the proposal eat into current browsers under the hood behaviour that can't be polyfilled — eg custom handling of 'static' resources via the DOM such as img etc. But as a replacement for XHR, it's delicious :)

**Comment by trueadm** [comment]
Hey @barneycarroll,

Inferno is a very "modern" approach to virtual DOM. It aims to strip away static elements that generally plague typical virtual DOM implementations (diffing static vnodes is useless, so why do it?). It aims to return barebones from a render() call that is optimised for JIT compilers. Furthermore, it understands hot spots in the DOM and attempts to recycle nodes that can be consuming to construct and set (nodes with textNodes as a single child). It's still in under early development, I'm working closely with many of the people in this thread to improve it, as I with their implementations.

Like I mentioned earlier, I can add you to the Skype group if you're interested in hearing/discussing all things vdom. I notice you're not far from me either :) good to have more Brits involved in this scene.

**Comment by tailhook** [comment]
@trueadm, Hi, I'm interested in inferno and virtual DOM too. In particular I'm building a template engine[1] which currently compiles to cito.js, and that may benefit of your project too. My skype is PaulColomiets.

[1]: https://github.com/tailhook/marafet

**Comment by paldepind** [comment]
@lhorie @tailhook 

> > lifecycle methods, one thing that has been asked a couple of times for Mithril is the ability to handle pre-detach animations (or, in a nutshell, asynchronous detachments)

> I'm pretty sure I've seen a virtual dom library that done that. I don't remember which one. And I think that its very interesting thing to have out of the box without any ugly things on application side.

Unless I'm misunderstanding something then [Snabbdom](https://github.com/paldepind/snabbdom) has this lifecycle hook. Listeners of the hook receive the vnode that is about to be removed and a callback that they must invoke once they have done their job. Only when all registered listeners have invoked their callback is the element removed from the DOM.

**Comment by Bobris** [comment]
Removing animation in Bobril it could be solved also client side (in other VDom implementation probably in similar style).
See http://bobris.github.io/Bobril/router2/index.html#/planet/Earth opacity animation is done by JS and router actually does not know about it at all.

**Comment by paldepind** [comment]
I'm late to the talk here. But let me just share my thinking. Maybe it will be of use.

If I understand correctly the goal of the ideas here is to mostly get rid of the virtual DOM creation and diffing and instead replace it with a template language that looks similar to what people use today to create a virtual DOM tree but instead of creating that it creates a data structure which the library can consume once to figure out what the DOM should look like and where any dynamic values should be inserted.

To me that sounds more like a Vue/Ember approach than it sounds like virtual DOM (nothing wrong with that). But then how do you figure out when the dynamic values has changed? Vue uses observables faked with getters and setters and Ember uses value diffing. Neither of these seems ideal to me. They both rely on imperative techniques and my goal with virtual DOM is to create write more declarative code.

This reminds me of an experiment I did a while back. Functional reactive programming has a concept of reactive values. That is, values that can change over time. I.e. FRP gives us exactly a representation for changing values that we can use in a view library to figure out what is static and what is dynamic. In something called [flyview](https://github.com/paldepind/flyview) I paired my [FRP library](https://github.com/paldepind/flyd) with a hyperscript like language. Every changing value had to be a stream and the view library could then build the DOM only once and from then on simply observe the streams. The result was something that looked a lot like virtual DOM but was implemented more like a library based on observables (like Knockout). I still had to do children reconciliation but it completely avoided all other diffing and virtual DOM creation as well.

But there is another solution to the problems listed by @trueadm in [this comment](https://github.com/joelrich/citojs/issues/18#issuecomment-124695541): immutable data and thunks. That is an approach inspired by Om. If we treat JavaScript objects and arrays as if they where immutable (working on them with a functional library like [Ramda](https://github.com/ramda/ramda/) we can very easily cut out huge chunks of the virtual DOM creation process _and_ the diffing process.

Using Snabbdom and its thunk module we can do something like this:

```js
const numberInSpan = (n) => h('span', 'Number is ' + n)
[... later inside the vdom creation ...]
thunk('num', numberInSpan, 22)
```

In the above example the virtual DOM for the span will only be created and take part in the diffing process in the case that the number has actually changed. When we use immutable data structures we know that if we compare a previously seen object with a new object and they are strictly equal (`obj1 === obj2`) then nothing inside it can have changed. This is just comparison of pointers which is as cheap as comparing numbers. Thus we can extend the power in the above example to deep and nested structures.

virtual-dom has thunks as well and I am sure that you people here are familiar with the idea. The main obstacle that remains is how to actually write complex applications using immutable data. I've been making such experiments in [this GitHub repository](https://github.com/paldepind/noname-functional-frontend-framework/) borrowing a lot of ideas from the Elm programming language. With the architecture explored in the repository there is no mutations in application code and thus applying the thunk optimization is trivial. Among other things there is an implementation of the familiar todo application.

Now, using immutable data results in garbage being created but in this case the amount of garbage is proportional to the number of changes in state and not to the size of the virtual DOM tree. That seems much more favorable to me.

**Comment by joelrich** [comment]
@paldepind yes, my idea is to get rid of the virtual dom diffing and instead just diff values. I have thought about streams as well, but came to the conclusion that it would unlikely be faster. However, by carefully choosing the API you could probably switch the implementation from diff to streams or even observing.

Your response would have deserved a longer response, but I want to concentrate on coding right now :-)

**Comment by joelrich** [comment]
Quick update: The Cito rewrite does now pass the vdom benchmark test (`&test=true` parameter) and is by far the fastest library for updates. Initial rendering is about the same, but can be accelerated by recycling nodes. With the recycling enabled, it's 1/3 faster than the current Cito overall and there are still optimisations pending.

**Comment by tinchoz49** [comment]
Really nice @joelrich ! One question...

When you finish the rewrite of cito... are you planning to make a component framework on top of that?

**Comment by joelrich** [comment]
@tinchoz49 I can't promise it right now, but I will make sure that it provides everything which is required to build a component framework on top of it.

**Comment by paldepind** [comment]
Is it coming up on GitHub?

**Comment by joelrich** [comment]
@paldepind yes, I will push it to a branch once it's complete enough to be used for general use cases. At the moment, there is still important functionality missing.

**Comment by mneumann** [comment]
@joelrich Your new idea sounds very similar to the ```children: function(oldChildren)``` approach in current cito, isn't it? Is it "just" a slightly different API to make these things easier or is there a fundamental difference?

**Comment by joelrich** [comment]
@mneumann it is similar, but for a different purpose. Let's have a look at the following example:

```javascript
// Define template
var myHeader = cito.template(function (titleVar) {
	tag: 'h1',
	attrs: {'class': 'my-header'}
	text: titleVar.use(title => title.toUpperCase())
});

// Use template
var fragment = cito.appendTo(document.body); // Append placeholder
fragment.render(myHeader('Title 1')); // Create dom
fragment.render(myHeader('Title 2')); // Update dom
fragment.remove(); // Remove dom
```

Even though `myHeader` is called twice, the callback which is passed to `cito.template` is only called once to initialise the template. Since the expression (title.toUpperCase) is wrapped with a lambda, it can derive a dom create and update function from the template.

This create and update function is more efficient than virtual dom diffing because it does not have to create a virtual dom tree, does not have to normalise it, compares only the values and can leverage several optimisations like cloneNode and recycling which cannot (easily) be used in normal virtual dom libraries.

The main question for me is whether or not people will be willing to adopt this new style. What do you think about it?

**Comment by venning** [comment]
@joelrich, does this imply that all attributes/properties/children that may change must be identified as such by `.use()`?  This makes a lot of sense, but I am very curious how this plays out in practice.

I like the idea of logically isolating changes to only the fragments you know are affected by a given event.  Seems like that could help with larger codebases.  I'm assuming you wouldn't be able to update multiple fragments simultaneously this way and would have to place those nodes inside of the same fragment just in case the DOM updated in subsequent frames.

**Comment by joelrich** [comment]
@venning almost all, but there will be a few shortcuts.

No transformation:
```javascript
cito.template(function (textVar) {
	return {
		tag: 'div',
		// No use() required because the value isn't transformed
		attrs: {title: textVar},
		children: textVar}
	};
);
```

With using collections:
```javascript
var t = cito.template;
t(function (listVar) {
	return {
		tag: 'ul',
		// Maybe I will allow to leave t() away here
		children: listVar.map(t(function (itemVar) {
			return {
				tag: 'div',
				children: itemVar
			};
		}))
	}
});
```

Your own utilities:
```javascript
var dateFormat = cito.utility(function (date) {
	return d.toGMTString();
});

// Which is basically the same as:

function dateFormat(dateVar) {
	return dateVar.use(function (data) {
		return d.toGMTString();
	});
}

cito.template(function (dateVar) {
	return {
		tag: 'div',
		children: dateFormat(dateVar)
	};
)
```

As you can see, the reason why I call it `use` instead of `map` is because `map` can be used to map every collection item (instead of the collection itself).

Feedback is very much welcome!

**Comment by venning** [comment]
I really like what you're doing here.  Okay, my thoughts:

- Thinking back through your earlier comments, I'm understanding that cito is tracking/diffing *values* before deciding to execute templates.  This makes a lot more sense now, and is really what Angular, et al has been striving towards for so long.  I like it.

- The fact that `attrs: {title: textVar}` works would tempt me into thinking that `attrs: {title: textVar.toUpperCase()}` should work which, as I understand it, is incorrect.  I could see this being a gotcha.  Enforcing uniform access to underlying values would help with clarity when using cito.  Clarity is a huge selling point in this space; Angular has burned a lot of us.  In Mithril's version of "properties", this is addressed by forcing property values to be accessed by zero-argument function calls.  I'd have to see the rest of cito play out, but maybe something like `attrs: {title: textVar()}` in the first example would make it easier to use?

- I understand that there's [a lot of hate](https://www.reindex.io/blog/you-might-not-need-underscore/) regarding re-writing builtins like `array.map(fn)`, but I really get a lot of mileage out of using `_.map(arr, fn)`.  Specifically, I often allow `null`/`undefined` as valid inputs and functional equivalents to `[]`, making `arr.map` less appealing than `_.map`.  This avoids the clutter of defensive code.  As such, it might be cleaner and more helpful to have something like `cito.map(arr, fn)` that understands `fn` is a cito template and handles `null`/`undefined` silently.

- Do you intend on integrating events and handlers into this re-write, or do you intend that be to left to components framework(s) written on top of cito?

**Comment by venning** [comment]
To be clear, I would imagine that in the `attrs: {title: textVar()}` example, the call `textVar()` does **not** return the underlying value but rather whatever object cito is using to track changes within the template. ... Now that I write that, I'm not sure my suggestion makes sense.

**Comment by trueadm** [comment]
This is something I specifically avoided and told Joel he should avoid – specifically the `map` function. 

It's utterly confusing to the end-user when you start creating replica map functions; especially when you get to UTing a large application, you *expect* things to work one way and only that way. I think there's been a big shift towards this pattern.

I guess the point is: what's more important, a few extra milliseconds of rendering time or a couple of hours of development time? We all took a step back from Angular/Ember etc and embraced React and virtual DOM and it's ability to give the developer much more freedom in how they can operate (which has saved countless hours in development time) – I really don't want to see us drift back into that.

**Comment by venning** [comment]
@trueadm, I'm not actually sure what you're referring to, could you clarify?

---

Regarding this original example:
```javascript
var t = cito.template;
t(function (listVar) {
    return {
        tag: 'ul',
        // Maybe I will allow to leave t() away here
        children: listVar.map(t(function (itemVar) {
            return {
                tag: 'li',
                children: itemVar
            };
        }))
    }
});
```
I was assuming that `listVar` was a raw Javascript array but, looking at it again, @joelrich might already be doing what I was asking: is `listVar` something created by cito that comes with `.use` and `.map` methods?  If so, please disregard my comment about `cito.map`.

**Comment by trueadm** [comment]
@venning it should be now. Once upon a time it wasn't.

**Comment by joelrich** [comment]
@venning here my response to your thoughts:
* To be more precise, it compares the values before it updates the dom. In addition to this, it can skip the transformation of primitive values or immutable data structures if they haven't changed. Therefore, if you have an immutable list of immutable items, it can for example automatically skip diffing all the items if the list hasn't changed.
* That you cannot use `attrs: {title: textVar.toUpperCase()}` is unfortunate, but the only way to support this would be to use ES6 proxies (might be an option in a few years) or by rewriting the JS code (not acceptable for me). I think the only real option I have is to clearly document this limitation and try to detect usage mistakes in the dev version of Cito. Moreover, this obviously something which has to be described at the very beginning of the documentation. Last but not least, it is possible to avoid `use`/`map` entirely by transforming the values outside of the template.
* I wouldn't rewrite `map`. The arguments which are passed to the template callbacks are of a special type which has a few methods like `map`, `use` and maybe `get`. Without `map`, you would have to write the following for iterating over an array:

```javascript
var t = cito.template;
t(function (listVar) {
	return {
		tag: 'ul',
		children: listVar.use(list => list.map(t(function (listVar) {
			return {tag: 'li', children: listVar}
		})))
	}
})
```

Therefore, `listVar.map(callback)` is mainly a shortcut for `listVar.use(item => item.map(callback))` and the execution of it behaves exactly the same as `[].map` itself. At the moment, I do not really see a problem with it (apart from the fact that the arguments concept itself is something new and has to be understood first).
* Events, lifecycle callbacks and state will be supported.

@trueadm As I tried to explain above, `map` will not have to be used. Here are a few examples which do (almost) exactly the same:

ES5 array map:
```javascript
var t = cito.template;
var ul = t(function (childrenVar) {
	return {tag: 'ul', children: childrenVar};
});
var li = t(function (itemVar) {
	return {tag: 'li', children: itemVar};
});

// ...

fragment.render(ul(list.map(li)));
```

Internal variable type use + ES5 array map:
```javascript
var t = cito.template;
t(function (listVar) {
	return {
		tag: 'ul',
		children: listVar.use(list => list.map(t(function (listVar) {
			return {tag: 'li', children: listVar}
		})))
	}
})
```

Internal variable type map:
```javascript
var t = cito.template;
t(function (listVar) {
	return {
		tag: 'ul',
		children: listVar.map(t(function (listVar) {
			return {tag: 'li', children: listVar}
		}))
	}
})
```

Internal variable type mapTemplate (just another idea for now):
```javascript
cito.template(function (listVar) {
	return {
		tag: 'ul',
		children: listVar.mapTemplate(function (listVar) {
			return {tag: 'li', children: listVar}
		})
	}
})
```

**Comment by trueadm** [comment]
I think people genuinely like and enjoy the fact that they can now write expressive JavaScript without the need of a framework/library. It's become almost an anti-pattern now to use jQuery, Underscore, Lodash etc when you have the exact equivalent (and more) available in native code.

Why is this important? For testing, extending and re-usability. It's very easy to mock the native prototype and to extend upon it without actually changing it. Furthermore, it's predictable and thus can be highly optimised by browser vendors.

Plus you can use iterables on many of the different native operations available now and inject them into native Maps and WeakMaps. It's great that vdom frameworks have allowed someone to do this:

peopleList.reverse().reduce(reduction).from(otherPeople).of(collection).map(renderPerson)

That's 100% re-usable, testable and scalable. I can take that same code and run it independently without a framework (as it's based on a data model) and expect the same result every time. 

When you start tightly coupling your data model and your templating schema, you make for a very uncomfortable development experience where, what should take a few days to refactor, can take weeks or months.

**Comment by venning** [comment]
Thank you @joelrich for the thoroughness of your response.  I hope I'm not distracting you from development.  This all makes a lot of sense.  The `.mapTemplate` idea was exactly what I was thinking.

**Comment by tailhook** [comment]
> The main question for me is whether or not people will be willing to adopt this new style. What do you think about it?

I personally don't believe in writing cito.js as is, and believe in transpiling. So the syntax is ok.

What I'm worry about is what's about lists. If they are not immutable, what will you compare, objects in the list or virtual dom generated? Comparing objects is bad in JS, because of bad equals operator. Comparing virtual dom nodes generated for list items looks like, you need to generate whole virtual dom in depth anyway, like in current cito, except you optimize the only single "root" template which is static.

**Comment by joelrich** [comment]
@tailhook even for mutable arrays it does not compare the virtual dom, but only the values. This works because each list item is also rendered with a template function. The trick is that template functions return a very slim and flat object which only contains the values instead of the whole virtual dom tree.

**Comment by SlinSo** [comment]
Hi @joelrich, @lhorie , @trueadm , ...

are you guys combining your efforts to create one virtual-dom library supported by more than one main delevoper? at the moment the vdom libraries seem very framented to me and everyone is working on performance improvements borrowing ideas from other libs. nothing bad, and for personal projects i'm totally fine, but it would be really nice to get a library like citojs oder mithriljs supported and developed together by some of you guys.

are there any plan on combining efforts other than exchanging ideas and concepts?

**Comment by mneumann** [comment]
@joelrich Your idea makes perfect sense and I am eagerly waiting to play with the code! I think the best way to determine if it works out well is to actually use it :)

Will cito compare the values itself for equality, or will there be a hook so that I can use my own comparison function? I am not using immutabilty, instead I just use a modification counter. Every modification to the object increases the count by one. Then in cito I only have to compare the counter value to the old counter value (stored in the vdom) and the reference to the object to see if I have to redraw the subtree. Or in case of immutability, will cito make it easy to replace the template variable with a new value? Will it just be ```itemVar.set(newVal)```?

**Comment by joelrich** [comment]
@SlinSo on the one hand, it would make a lot of sense if we all worked together on a common library. On the other hand, my impression is that we (or at least I) aren't ready for this yet because we are still experimenting with fundamentals which have a big influence on the API.

If the goal was "just" to implement a great virtual dom library, then we could start with it now because we have learnt enough about what the difficulties are and what users expect from a virtual dom library. However, my goal isn't necessary to implement a virtual dom library, but to implement a flexible template engine which performs extremely well. Although this project started as a virtual dom library, I came to the conclusion that there might be better concepts to reach this goal - and I am experimenting with this now.

In the short term, the situation does not look ideal for users; I agree. However, the "competition" which we have right now has already led to improvements of libraries which are now used in production (see for example https://github.com/lhorie/mithril.js/issues/712) and leads to improved and new concepts which might have an influence on how template engines will work in the future.

Nevertheless, at some point we will have experimented enough and have gotten enough feedback to decide how to continue and at this point we should also try to intensify our collaboration.

This is probably not what you wanted to hear, but I hope it gives you at least an insight why I am working on my own project instead of contributing to an existing one.

**Comment by joelrich** [comment]
@mneumann I have thought about this as well. This would bascially be the equivalent of `shouldComponentUpdate` on a variable level. My idea is that one will be able to add type information to the variables. Example:

```javascript
var counterListType = {
	equals: function (a, b) {
		return a.__counter === b.__counter;
	},
	customMap: function (list, callback) {
		return list.__values.map(callback);
	}
};

var t = cito.template;
t(function (listVar) {
    return {
        tag: 'ul',
        children: listVar.is(counterListType).map(t(function (itemVar) {
            return {tag: 'li', children: itemVar}
        }))
    }
})
```

The alternative would be to detect value types automatically, but this has the potential to slow everything down.

By the way, once there are annotions in JavaScript, this could look much nicer:

```javascript
@Template
function list(@Type(counterList) listVar) {
	// ...
}
```

**Comment by tailhook** [comment]
> The alternative would be to detect value types automatically, but this has the potential to slow everything down.

Could TypeScript or Flow help with it?

**Comment by joelrich** [comment]
@tailhook if the type hints were retained in the transpiled source code, then it would help. I don't think this is possible right now, but it would make sense to talk to the TypeScript/Flow team to discuss such a feature.

**Comment by tailhook** [comment]
> @tailhook if the type hints were retained in the transpiled source code, then it would help.

AFAIU, it's easy to do with babel plugin.

**Comment by joelrich** [comment]
@tailhook do you have a link to an example/documentation?

**Comment by tailhook** [comment]
@joelrich http://babeljs.io/docs/advanced/plugins/ ?

Also according to [roadmap] all transformers will be plugins.

[roadmap]: https://github.com/babel/babel/issues/2168

**Comment by joelrich** [comment]
@tailhook what I would need is the capability of the transpiler to retain the type information in a way which makes it easy to access it during execution. The following TypeScript function

```typescript
function add(a: number, b: number): number {
	return a + b;
}
```

would for example have to be transpiled to something like

```javascript
function add(a, b) {
	return a + b;
}
add.__signature = {parameterTypes: ['number', 'number'], returnType: 'number'};
```

Since you wouldn't want to retain all signatures, the question is how to enable such a feature selectively.

**Comment by tailhook** [comment]
@joelrich sure, you may create a plugin, which executed before the "flow" plugin (that strips types) and translates the code into the following:
```
function add(a: number, b: number) :number {
    return a + b;
}
add.__signature = {parameterTypes: ['number', 'number'], returnType: 'number'};
```

I.e. add additional code, but keep code valid for flow. (It's actually done on AST level not text).

Or you could use flow syntax for your own annotations and strip them yourself, but that's probably bad idea because doesn't allow to use flow for type-checking.

**Comment by mneumann** [comment]
@joelrich of course it would be nice to automate this, but in my case I am totally fine to write the signature by hand, so IMHO this shouldn't be an issue.

**Comment by joelrich** [comment]
@tailhook sure, I could write such a plugin, but I don't think it's worth the effort. Moreover, I think it would make more sense to add this capability to the TypeScript/Flow transpiler itself. There are likely other use cases for it.

@mneumann agreed :-)

**Comment by ricochet1k** [comment]
@joelrich I love the design for the new version of cito. Any updates? Do you have any code for it available?

**Comment by joelrich** [comment]
@ricochet1k I wished to have an update, but unfortunately I have barely time to work on this project at the moment because I have started a part-time school which is more time-consuming than I expected. I hope that I will find time to work on it during my school vacation though.

**Comment by joelrich** [comment]
I am sorry to say that I don't manage to continue with the rewrite under the current circumstances. I simply don't have enough time for it and I cannot afford to fail the exam next August.

For the current code base, I will continue fixing bugs when they appear.

If anybody who is working on a similar library is interested in my findings, please send me an email and I will share them with you.

**Comment by stevenvachon** [comment]
[InfernoJS](https://github.com/trueadm/inferno) is faster.

**Comment by kuraga** [comment]
@joelrich it would be nice you make decision about Promises and tag+publish new version...

**Comment by joelrich** [comment]
@kuraga My plan was to add a utility for promises. This is necessary because without you cannot render a loading/pending message. A relatively inefficient version could look like this:

```javascript
function p(promise, pending, rejected) {
    var node = {children: pending};
    promise.then(function (newNode) {
        cito.vdom.update(node, {children: newNode});
    }, function (newNode) {
        cito.vdom.update(node, {children: rejected || newNode});
    });
    return node;
}

// Usage example

var promise = new Promise((resolve, reject) => {
    setTimeout(() => {
        resolve({tag: 'h1', children: 'Header'});
    }, 500);
});

var node = cito.vdom.append(document.body, {
    tag: 'div',
    children: p(promise, 'Loading', 'Rejected')
});

```

For most cases this should be sufficient. Unfortunately, I cannot afford to implement a more efficient version with test coverage at the moment.

**Comment by kuraga** [comment]
Ok then just tag and publish, please :smile:
